### PR TITLE
feat: Antrea plugin support in AKS Engine

### PR DESCRIPTION
**Pull request by reachjainrahul**
**Reason for Change**:

Add Antrea networking and policy plugin in AKS engine.  https://github.com/vmware-tanzu/antrea

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ X] includes documentation
- [ X] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: parts/k8s/addons/kubernetesmasteraddons-antrea-daemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,495 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (jackfrancis) [inline parts/k8s/addons/kubernetesmasteraddons-antrea-daemonset.yaml]: Some backstory: the `parts/k8s/addons/` directory is a "legacy" addons implementation. This is not obvious. :/

So, let's move this spec under `parts/k8s/containeraddons/`, which will allow us to enable much more flexible user configuration.

Comment (reachjainrahul) [reply parts/k8s/addons/kubernetesmasteraddons-antrea-daemonset.yaml]: Ok.. will move the YAML file under containeraddons dir.  Thanks

File: parts/k8s/containeraddons/kubernetesmasteraddons-antrea-daemonset.yaml [hunk 2]
```diff
@@ -0,0 +1,495 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (jianjuns) [inline parts/k8s/containeraddons/kubernetesmasteraddons-antrea-daemonset.yaml]: Actually this yaml creates a Deployment, a DaemonSet, and multiple other resources. Should we name it antrea-deployment.yaml instead?

Comment (reachjainrahul) [reply parts/k8s/containeraddons/kubernetesmasteraddons-antrea-daemonset.yaml]: Most plugin use this convention. I named it accordingly.

File: docs/tutorials/deploy.md [hunk 3]
```diff
@@ -141,6 +141,7 @@ aks-engine generate --set agentPoolProfiles[0].count=5,agentPoolProfiles[1].name
 
 * To enable the optional network policy enforcement using calico, you have to set the parameter during this step according to this [guide](../topics/features.md#optional-enable-network-policy-enforcement-using-calico)
 * To enable the optional network policy enforcement using cilium, you have to set the parameter during this step according to this [guide](../topics/features.md#optional-enable-network-policy-enforcement-using-cilium)
+* To enable the optional network policy enforcement using antrea, you have to set the parameter during this step according to this [guide](../topics/features.md#optional-enable-network-policy-enforcement-using-antrea)
```
[the remark this answers is not in the document]
Comment (reachjainrahul) [reply docs/tutorials/deploy.md]: This section is for policy mode only. For networking there is separate doc.

File: examples/networkpolicy/kubernetes-antrea.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorRelease": "1.13",
```
[the remark this answers is not in the document]
Comment (reachjainrahul) [reply examples/networkpolicy/kubernetes-antrea.json]: Not really. This is just a reference example.

File: examples/networkpolicy/README.md [hunk 5]
```diff
@@ -97,3 +98,20 @@ The kubernetes-cilium deployment template enables Cilium networking and policies
...
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "kubernetesConfig": {
+        "networkPolicy": "antrea"
+      }
+```
+
+### Post installation
+
+For the latest documentation on Antrea, please refer to [this](https://github.com/vmware-tanzu/antrea)
```
[the remark this answers is not in the document]
Comment (reachjainrahul) [reply examples/networkpolicy/README.md]: Will update it with yaml 0.2.0 checkin.

**Comment by jackfrancis** [comment]
Hello again, I've been enumerating through the existing addons and in my refactor of this one it appears we're in a similar functional space as antrea:

https://github.com/Azure/aks-engine/pull/2480

Specifically, it appears that like the existing cilium implementation, antrea delivers both IPAM and a NetworkPolicy functionality, and those will be represented by `"networkPlugin": "antrea"` and `"networkPolicy": "antrea"`, respectively. Also, like cilium it appears that the antrea IPAM implementation depends upon artifacts delivered via a daemonset in the addon spec.

Is that correct?

If so, I would advocate that we deliver the antrea specification as follows:

- if the user specifies `"networkPolicy": "antrea"` in the api model, we automatically set the `"networkPlugin": "antrea"` configuration
- if the user specifies only ``"networkPlugin": "antrea"``, we throw a validation error at ARM template generation time

In other words, we want to always deliver a cluster configuration that looks like:

```
...
    "networkPolicy": "antrea",
    "networkPlugin": "antrea
...
```

And we permit the user only specifying the networkPolicy part, automatically filling in the antrea networkPlugin configuration for him/her. But we don't permit *just* an antrea networkPlugin cluster configuration, because that may mislead the user into thinking that we support *just* an antrea CNI implementation for IPAM without NetworkPolicy, which is in fact not true.

Does that all sound right?

**Comment by jianjuns** [comment]
@jackfrancis your suggestion makes sense to me. We do not support a model to use Antrea for IPAM and connectivity only, but use other CNI for NetworkPolicy.

**Comment by reachjainrahul** [comment]
> @jackfrancis your suggestion makes sense to me. We do not support a model to use Antrea for IPAM and connectivity only, but use other CNI for NetworkPolicy.

**Comment by reachjainrahul** [comment]
> Hello again, I've been enumerating through the existing addons and in my refactor of this one it appears we're in a similar functional space as antrea:
> 
> #2480
> 
> Specifically, it appears that like the existing cilium implementation, antrea delivers both IPAM and a NetworkPolicy functionality, and those will be represented by `"networkPlugin": "antrea"` and `"networkPolicy": "antrea"`, respectively. Also, like cilium it appears that the antrea IPAM implementation depends upon artifacts delivered via a daemonset in the addon spec.
> 
> Is that correct?
> 
> If so, I would advocate that we deliver the antrea specification as follows:
> 
> * if the user specifies `"networkPolicy": "antrea"` in the api model, we automatically set the `"networkPlugin": "antrea"` configuration
> * if the user specifies only `"networkPlugin": "antrea"`, we throw a validation error at ARM template generation time
> 
> In other words, we want to always deliver a cluster configuration that looks like:
> 
> ```
> ...
>     "networkPolicy": "antrea",
>     "networkPlugin": "antrea
> ...
> ```
> 
> And we permit the user only specifying the networkPolicy part, automatically filling in the antrea networkPlugin configuration for him/her. But we don't permit _just_ an antrea networkPlugin cluster configuration, because that may mislead the user into thinking that we support _just_ an antrea CNI implementation for IPAM without NetworkPolicy, which is in fact not true.
> 
> Does that all sound right?

@jackfrancis Would you update the patch for validation error? Your suggestion sounds good to me too.

File: docs/tutorials/deploy.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhiraut) [reply docs/tutorials/deploy.md]: correct.. what i was looking for was a way to ensure that we do not convey the message that antrea can be used in a policy only with other networking cni and that there should be some validation.. Jack answered that in his comment regarding validation. so we should be good

[the remark this answers is not in the document]
Comment (reachjainrahul) [reply docs/tutorials/deploy.md]: actually 3 combination were added in the validation.

1) Just NetworkPlugin, Without any Policy. If any policy is specified other than antrea, its a validation error
2) Just NetworkPolicy. Without any NetworkPlugin. If network plugin is specified other than antrea, its a validation error
3) Both Network and Policy plugin as antrea.

We would remove option 1.  Hope it make sense.

**Comment by jackfrancis** [comment]
Hi @reachjainrahul and folks,

I've rebased this PR and done a bit of cleanup so it resembles other addons (specifically cilium). I have a passing E2E test, so I think this is ready to merge. However, I am seeing this on my cluster:

```
$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                            READY   STATUS             RESTARTS   AGE
...
kube-system   antrea-agent-m526s                              2/2     Running            1          45m
kube-system   antrea-agent-qjrk8                              2/2     Running            0          45m
kube-system   antrea-agent-zg5c4                              2/2     Running            0          45m
kube-system   antrea-controller-6795f5bbfd-4b6qr              1/1     Running            0          45m
...
kube-system   kube-controller-manager-k8s-master-38231306-0   0/1     CrashLoopBackOff   8          45m
...
kube-system   kube-scheduler-k8s-master-38231306-0            0/1     CrashLoopBackOff   8          45m

```

Note the 8 restarts for both the controller-manager and scheduler pods. Both seem to be exhibiting a similar failure symptom:

```
$ kubectl logs kube-controller-manager-k8s-master-38231306-0 -n kube-system
...
I1219 23:28:02.459750       1 leaderelection.go:281] failed to renew lease kube-system/kube-controller-manager: failed to tryAcquireOrRenew context deadline exceeded
F1219 23:28:02.459832       1 controllermanager.go:281] leaderelection lost
```

```
$ kubectl logs kube-scheduler-k8s-master-38231306-0 -n kube-system
...
I1219 23:28:04.120899       1 leaderelection.go:281] failed to renew lease kube-system/kube-scheduler: failed to tryAcquireOrRenew context deadline exceeded
E1219 23:28:04.121324       1 server.go:254] lost master
lost lease
```

Have you seen this in any of your tests? As you can imagine, pod scheduling/reconciliation is significantly de-optimized on a cluster like this, as the key componentry responsible for that is regularly offline.

File: parts/k8s/containeraddons/kubernetesmasteraddons-antrea-daemonset.yaml (hunk 2 above)
Comment (jackfrancis) [reply parts/k8s/containeraddons/kubernetesmasteraddons-antrea-daemonset.yaml]: I updated it to just `"antrea.yaml"`. The convention you observed is a poor one, unfortunately, but totally understood why you were following the pre-existing patterns!

File: docs/tutorials/deploy.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jackfrancis) [reply docs/tutorials/deploy.md]: I removed option 1 as a valid option following the recent rationalization of cilium, which follows a similar pattern.

File: parts/k8s/cloud-init/artifacts/cse_config.sh [hunk 6]
```diff
@@ -312,6 +312,11 @@ ensureKubelet() {
         sleep 3
     done
     {{end}}
+    {{if HasAntreaNetworkPolicy}}
+    while [ ! -f /etc/cni/net.d/10-antrea.conf ]; do
```
Comment (jackfrancis) [inline parts/k8s/cloud-init/artifacts/cse_config.sh]: Note: if the daemonset implementation ever changes the name lf this CNI config file, we'll have to update this file wait implementation.

File: parts/k8s/containeraddons/antrea.yaml [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+        - mountPath: /host/proc
+          name: host-proc
+          readOnly: true
+        - mountPath: /host/var/run/netns
+          mountPropagation: HostToContainer
+          name: host-var-run-netns
+          readOnly: true
+      - command:
+        - start_ovs
+        image: antrea/antrea-ubuntu:latest
```
Comment (abhiraut) [inline parts/k8s/containeraddons/antrea.yaml]: we should perhaps pin this to a version and not latest? antrea/antrea-ubuntu:v0.2.0

Comment (jackfrancis) [reply parts/k8s/containeraddons/antrea.yaml]: Agreed an immutable reference would be preferable. Doesn't have to block merge, though, at this phase.

Also, the references to the images are here:

https://github.com/Azure/aks-engine/pull/2407/files#diff-bb9b6cb6f08a63800be959ba49eaf714R42

File: test/e2e/test_cluster_configs/network_policy/antrea.json [hunk 8]
```diff
@@ -0,0 +1,46 @@
+{
+	"env": {
+	},
+	"options": {
+		"allowedOrchestratorVersions": ["1.13", "1.14", "1.15", "1.16"]
```
Comment (jackfrancis) [inline test/e2e/test_cluster_configs/network_policy/antrea.json]: @reachjainrahul do @abhiraut do we know for sure that antrea doesn't work w/ k8s 1.17?

Comment (abhiraut) [reply test/e2e/test_cluster_configs/network_policy/antrea.json]: in theory yes, it should work.. however let us update this only after we run some ci tests against 1.17? @jianjuns any opinion?

File: pkg/engine/params_k8s.go [hunk 9]
```diff
@@ -122,6 +121,7 @@ func assignKubernetesParameters(properties *api.Properties, parametersMap params
 				// Kubernetes node binaries as packaged by upstream kubernetes
 				// example at https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.11.md#node-binaries-1
 				addValue(parametersMap, "windowsKubeBinariesURL", kubernetesConfig.WindowsNodeBinariesURL)
+				addValue(parametersMap, "kubeServiceCidr", kubernetesConfig.ServiceCIDR)
```
Comment (reachjainrahul) [inline pkg/engine/params_k8s.go]: Why this revert? Antrea requires ServiceCIDR. Adding here will set it for Windows VM only, right ???

File: pkg/engine/template_generator.go [hunk 10]
```diff
@@ -577,8 +577,8 @@ func getContainerServiceFuncMap(cs *api.ContainerService) template.FuncMap {
 		"HasCiliumNetworkPolicy": func() bool {
 			return cs.Properties.OrchestratorProfile.KubernetesConfig.NetworkPolicy == NetworkPolicyCilium
 		},
-		"HasAntreaNetworkPlugin": func() bool {
-			return cs.Properties.OrchestratorProfile.KubernetesConfig.NetworkPlugin == NetworkPluginAntrea
+		"HasAntreaNetworkPolicy": func() bool {
+			return cs.Properties.OrchestratorProfile.KubernetesConfig.NetworkPlugin == NetworkPolicyAntrea
```
Comment (reachjainrahul) [inline pkg/engine/template_generator.go]: Shouldnt this be NetworkPolicy check ?

File: pkg/engine/transform/transformtestfiles/k8s_agent_upgrade_template.json [hunk 11]
```diff
@@ -1166,7 +1159,7 @@
     "kubeConfigCertificate": "[parameters('kubeConfigCertificate')]",
     "kubeConfigPrivateKey": "[parameters('kubeConfigPrivateKey')]",
     "kubeDnsServiceIp": "10.0.0.10",
-    "kubeServiceCidr": "[parameters('kubeServiceCidr')]",
+    "kubeServiceCidr": "10.0.0.0/16",
```
Comment (reachjainrahul) [inline pkg/engine/transform/transformtestfiles/k8s_agent_upgrade_template.json]: Is this revert intentional?

File: pkg/engine/transform/transformtestfiles/k8s_slb_template.json [hunk 12]
```diff
@@ -508,12 +508,6 @@
         },
         "type": "string"
       },
-      "kubeServiceCidr": {
```
Comment (reachjainrahul) [inline pkg/engine/transform/transformtestfiles/k8s_slb_template.json]: Same for all json file. is this intentional?

File: pkg/api/k8s_versions.go [hunk 13]
```diff
@@ -39,6 +39,10 @@ const (
 	ciliumCleanStateImageReference                    string = "docker.io/cilium/cilium-init:2018-10-16"
 	ciliumOperatorImageReference                      string = "docker.io/cilium/operator:v1.4"
 	ciliumEtcdOperatorImageReference                  string = "docker.io/cilium/cilium-etcd-operator:v2.0.5"
+	antreaControllerImageReference                    string = "antrea/antrea-ubuntu:latest"
```
Comment (reachjainrahul) [inline pkg/api/k8s_versions.go]: Antrea 0.2.0 is released today. Could we please update this to antrea/antrea-ubuntu:v0.2.0

**Comment by reachjainrahul** [comment]
I didnt see this in my test which I ran 2 weeks ago. @jianjuns any known issue with latest antrea? my antrea yaml referred to master.

**Comment by reachjainrahul** [comment]
> Hi @reachjainrahul and folks,
> 
> I've rebased this PR and done a bit of cleanup so it resembles other addons (specifically cilium). I have a passing E2E test, so I think this is ready to merge. However, I am seeing this on my cluster:
> 
> ```
> $ kubectl get pods --all-namespaces
> NAMESPACE     NAME                                            READY   STATUS             RESTARTS   AGE
> ...
> kube-system   antrea-agent-m526s                              2/2     Running            1          45m
> kube-system   antrea-agent-qjrk8                              2/2     Running            0          45m
> kube-system   antrea-agent-zg5c4                              2/2     Running            0          45m
> kube-system   antrea-controller-6795f5bbfd-4b6qr              1/1     Running            0          45m
> ...
> kube-system   kube-controller-manager-k8s-master-38231306-0   0/1     CrashLoopBackOff   8          45m
> ...
> kube-system   kube-scheduler-k8s-master-38231306-0            0/1     CrashLoopBackOff   8          45m
> ```
> 
> Note the 8 restarts for both the controller-manager and scheduler pods. Both seem to be exhibiting a similar failure symptom:
> 
> ```
> $ kubectl logs kube-controller-manager-k8s-master-38231306-0 -n kube-system
> ...
> I1219 23:28:02.459750       1 leaderelection.go:281] failed to renew lease kube-system/kube-controller-manager: failed to tryAcquireOrRenew context deadline exceeded
> F1219 23:28:02.459832       1 controllermanager.go:281] leaderelection lost
> ```
> 
> ```
> $ kubectl logs kube-scheduler-k8s-master-38231306-0 -n kube-system
> ...
> I1219 23:28:04.120899       1 leaderelection.go:281] failed to renew lease kube-system/kube-scheduler: failed to tryAcquireOrRenew context deadline exceeded
> E1219 23:28:04.121324       1 server.go:254] lost master
> lost lease
> ```
> 
> Have you seen this in any of your tests? As you can imagine, pod scheduling/reconciliation is significantly de-optimized on a cluster like this, as the key componentry responsible for that is regularly offline.

I ran E2E test with Antrea 0.2.0 and didnt see the issue. 


```
azureuser@k8s-master-41238993-0:~$ kubectl get pods -o wide -n kube-system 
NAME                                            READY   STATUS    RESTARTS   AGE   IP             NODE                                 NOMINATED NODE   READINESS GATES
antrea-agent-ct7fp                              2/2     Running   0          26m   10.240.255.5   k8s-master-41238993-0                <none>           <none>
antrea-agent-fclvn                              2/2     Running   1          26m   10.240.0.5     k8s-agentpool1-41238993-vmss000001   <none>           <none>
antrea-agent-zv74b                              2/2     Running   0          26m   10.240.0.4     k8s-agentpool1-41238993-vmss000000   <none>           <none>
antrea-controller-5b9b8d5bb9-pcllg              1/1     Running   0          26m   10.240.0.5     k8s-agentpool1-41238993-vmss000001   <none>           <none>
kube-addon-manager-k8s-master-41238993-0        1/1     Running   0          25m   10.240.255.5   k8s-master-41238993-0                <none>           <none>
kube-apiserver-k8s-master-41238993-0            1/1     Running   0          25m   10.240.255.5   k8s-master-41238993-0                <none>           <none>
kube-controller-manager-k8s-master-41238993-0   1/1     Running   0          25m   10.240.255.5   k8s-master-41238993-0                <none>           <none>
kube-scheduler-k8s-master-41238993-0            1/1     Running   0          25m   10.240.255.5   k8s-master-41238993-0                <none>           <none>
kubernetes-dashboard-7947fffdf5-q57cc           1/1     Running   0          26m   10.244.0.5     k8s-agentpool1-41238993-vmss000001   <none>           <none>
metrics-server-69b44566d5-fwftw                 1/1     Running   0          26m   10.244.0.4     k8s-agentpool1-41238993-vmss000001   <none>           <none>
```
